src/SecTester.Scan/Scan.cs

```diff
@@ -1,58 +1,164 @@
-  private readonly ScanState _scanState = new(ScanStatus.Pending);
+  private ScanState _scanState = new(ScanStatus.Pending);
+  private readonly SemaphoreSlim _scanStateSemaphore = new(1, 1);
+
+  public string Id { get; }
+
+  private bool InternalDone => DoneStatuses.Contains(_scanState.Status);
+  private bool InternalActive => ActiveStatuses.Contains(_scanState.Status);
 
-  public Scan(Scans scans, ScanOptions options)
+  public bool Active
   {
+    get
+    {
+      using var _ = SemaphoreLock.Lock(_scanStateSemaphore);
+      return InternalActive;
+    }
+  }
+
+  public bool Done
+  {
+    get
+    {
+      using var _ = SemaphoreLock.Lock(_scanStateSemaphore);
+      return InternalDone;
+    }
+  }
+
+  public Scan(string id, Scans scans, ILogger logger, ScanOptions options)
+  {
+    Id = id ?? throw new ArgumentNullException(nameof(id));
     _scans = scans ?? throw new ArgumentNullException(nameof(scans));
     _options = options ?? throw new ArgumentNullException(nameof(options));
+    _logger = logger ?? throw new ArgumentNullException(nameof(logger));
   }
 
-  public bool Active => ActiveStatuses.Contains(_scanState.Status);
+  public async ValueTask DisposeAsync()
+  {
+    try
+    {
+      await RefreshState().ConfigureAwait(false);
+
+      if (!Active)
+      {
+        await _scans.DeleteScan(Id).ConfigureAwait(false);
+      }
+    }
+    catch
+    {
+      // ignore
+    }
 
-  public bool Done => DoneStatuses.Contains(_scanState.Status);
+    _scanStateSemaphore.Dispose();
 
-  public void Dispose()
-  {
     GC.SuppressFinalize(this);
   }
 
-  public Task<IEnumerable<Issue>> Issues(CancellationToken cancellationToken = default)
+  public Task<IEnumerable<Issue>> Issues()
+  {
+    return _scans.ListIssues(Id);
+  }
+
+  public async Task Stop()
   {
-    throw new NotImplementedException();
+    try
+    {
+      await RefreshState().ConfigureAwait(false);
+
+      if (Active)
+      {
+        await _scans.StopScan(Id).ConfigureAwait(false);
+      }
+    }
+    catch
+    {
+      // ignore
+    }
   }
 
-  public Task Stop()
+  public async IAsyncEnumerable<ScanState> Status(
+    [EnumeratorCancellation] CancellationToken cancellationToken = default)
   {
-    throw new NotImplementedException();
+    while (Active)
+    {
+      await Task.Delay(_options.PollingInterval ?? DefaultPollingInterval, cancellationToken).ConfigureAwait(false);
+
+      yield return await RefreshState(cancellationToken).ConfigureAwait(false);
+    }
   }
 
-  public IAsyncEnumerable<ScanState> Status(CancellationToken cancellationToken = default)
+  private async Task<ScanState> RefreshState(CancellationToken cancellationToken = default)
   {
-    throw new NotImplementedException();
+    ScanState newState;
+
+    if (_scanStateSemaphore.CurrentCount == 0)
+    {
+      using var _ = await SemaphoreLock.LockAsync(_scanStateSemaphore, cancellationToken).ConfigureAwait(false);
+      newState = _scanState;
+    }
```

Comment:
Well, if `CurrentCount` is 0, this means the lock has already been acquired

Reply:
Changed to flag. A lock may be acquired by `Active` and `Done` getters.